### PR TITLE
refactor: move to type module

### LINKED ISSUE #152: 🐛 Bug: Can not import this in ESM package
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

No error with flat config

### Actual

```
TypeError: Cannot read properties of undefined (reading 'docs')
```

### Additional Info

Related: #91 

If you use this plugin like this:

```js
import * as pluginPackageJson from 'eslint-plugin-package-json'

{
  files: ['package.json'],
  plugins: {
    'package-json': pluginPackageJson,
  },
  rules: pluginPackageJson.configs.recommended.rules,
},
```

You will get an error, because flat config use esm. But in `lib/index.mjs`, it always use default import from a cjs file.

https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/blob/baf1aaa2696c3da4634f21f0d091669b3d3fb97d/src/index.ts#L1-L9

The imported `validVersion` looks like

```
{
  "default": {
    "meta": {
      "docs": {
        "category": "Best Practices",
        "description": "Enforce that package versions are valid semver specifiers",
        "recommended": true
      },
      "hasSuggestions": true,
      "messages": {
        "invalid": "Version is not a valid semver specifier.",
        "type": "\"version\" should be a string."
      }
    }
  }
}
```

So `rule.meta` is `undefined`

https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/blob/baf1aaa2696c3da4634f21f0d091669b3d3fb97d/src/index.ts#L26-L28

### PR DESCRIPTION
**Pull request by hyoban**
## PR Checklist

-   [x] Addresses an existing open issue: part of #91 
-   [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

This refactor allows us to use flat config now, although no default preset is provided yet. fix #152 

```js
import * as pluginPackageJson from 'eslint-plugin-package-json'
import parserJsonc from 'jsonc-eslint-parser'

export default [
  {
    files: ['package.json'],
    languageOptions: {
      parser: parserJsonc,
    },
    plugins: {
      'package-json': pluginPackageJson,
    },
    rules: pluginPackageJson.configs.recommended.rules,
  },
]
```

### CODE REVIEW

File: package.json
Comment (JoshuaKGoldberg) [inline package.json:1]: 🤔 I'm confused. How does switching this to a module change flat config support?

I realize #91 doesn't have much info, so I just posted some there: https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/91#issuecomment-1953437450.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks for getting started on this! 🙌

File: package.json
Comment (hyoban) [reply package.json:1]: This is related to the extension in import.

Currently, why we can not use this plugin in flat config(esm) is explained in #152. After moving to `type: module`, we can correctly import the esm build.

Comment (JoshuaKGoldberg) [reply package.json:1]: Got it, so the issue wasn't a duplicate after all. Will re-open. Thanks!

**Comment by hyoban** [comment]
I feel like using named imports is more of a compromise. We can now use `type : module` without dropping cjs support.

**Comment by JoshuaKGoldberg** [comment]
Yeah I see what you're saying - it makes sense that named exports is a bit of a compromise. And you're right that we could switch to `type: module` and consumers would still be able to work with it. But I worry about it 😅 and would rather do that as a breaking change down the line.

Maybe: let's go with `named exports` for now, and file switching to type module for a future version?

**Comment by hyoban** [comment]
Close this one for now

**Pull request closed without merging**